pkg/reconciler/reconciler.go

```diff
@@ -521,6 +521,19 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.
 		return ctrl.Result{}, err
 	}
 
+	// The finalizer must be present on the CR before we can do anything. Otherwise, if the reconciliation fails,
+	// there might be resources  created by the chart that will not be garbage-collected
+	// (cluster-scoped resources or resources in other namespaces, which are not bound by an owner reference).
+	// This is a safety measure to ensure that the chart is fully uninstalled before the CR is deleted.
+	if obj.GetDeletionTimestamp() == nil && !controllerutil.ContainsFinalizer(obj, uninstallFinalizer) {
+		log.V(1).Info("Adding uninstall finalizer")
+		obj.SetFinalizers(append(obj.GetFinalizers(), uninstallFinalizer))
+		if err := r.client.Update(ctx, obj); err != nil {
+			log.Error(err, "Failed to add uninstall finalizer")
+			return ctrl.Result{}, err
```

Comment:
IIRC, logging the error and then returning it results in a double log of the error because controller-runtime _also_ logs it.
```suggestion
			return ctrl.Result{}, fmt.Errorf("failed to add uninstall finalizer: %w", err)
```